api/cloud/oc_cloud_deregister.c

```diff
@@ -0,0 +1,242 @@
+                            bool useAccessToken)
+{
+  assert(p != NULL);
+
+  oc_cloud_context_t *ctx = p->ctx;
+  OC_DBG("try deregister device %zu by DELETE request", ctx->device);
+  oc_cloud_access_conf_t conf = {
+    .device = ctx->device,
+    .selected_identity_cred_id = ctx->selected_identity_cred_id,
+    .handler = cloud_deregistered_internal,
+    .user_data = p,
+    .timeout = timeout,
+  };
+  if (oc_string(ctx->store.ci_server) == NULL ||
+      conv_cloud_endpoint(ctx) != 0) {
+    goto error;
+  }
+  conf.endpoint = ctx->cloud_ep;
+
+  if (cloud_access_deregister(
+        conf, oc_string(ctx->store.uid),
+        useAccessToken ? oc_string(ctx->store.access_token) : NULL)) {
+    return 0;
+  }
+
+error:
+  cloud_set_last_error(ctx, CLOUD_ERROR_CONNECT);
+  return -1;
+}
+
+static void
+cloud_deregister_try_logged_in(oc_cloud_context_t *ctx,
+                               oc_cloud_status_t status, void *data)
+{
+  OC_DBG("Cloud deregister: logged in for device=%zu", ctx->device);
+  cloud_api_param_t *p = (cloud_api_param_t *)data;
+
+  if ((status & OC_CLOUD_LOGGED_IN) == 0) {
+    OC_ERR("Failed to login to cloud for deregister");
+    free_api_param(p);
+    cloud_context_clear(ctx);
+    return;
+  }
+
+  if (cloud_deregister_by_request(p, p->timeout, false) != 0) {
+    OC_ERR("Failed to deregister from cloud");
+    free_api_param(p);
+    cloud_context_clear(ctx);
+  }
+}
+
+static oc_event_callback_retval_t
+cloud_deregister_refreshed_token_async(void *data)
+{
+  cloud_api_param_t *p = (cloud_api_param_t *)data;
+  // short access token -> we can use it in query and deregister without login
+  if (check_accesstoken_for_deregister(p->ctx)) {
+    if (cloud_deregister_by_request(p, p->timeout, true) != 0) {
+      OC_ERR("Failed to deregister from cloud");
+      free_api_param(p);
+      cloud_context_clear(p->ctx);
+    }
+    return OC_EVENT_DONE;
+  }
+
+  // long access token -> we must login and then deregister without token
+  if (cloud_login(p->ctx, cloud_deregister_try_logged_in, p, p->timeout) != 0) {
+    OC_ERR("Failed to login to cloud for deregister");
+    free_api_param(p);
+    cloud_context_clear(p->ctx);
+    return OC_EVENT_DONE;
+  }
+  return OC_EVENT_DONE;
+}
+
+static void
+cloud_deregister_try_refreshed_token(oc_cloud_context_t *ctx,
+                                     oc_cloud_status_t status, void *data)
+{
+  OC_DBG("Cloud deregister: refreshed token for device=%zu", ctx->device);
+  if ((status & OC_CLOUD_REFRESHED_TOKEN) == 0) {
+    OC_ERR("Failed to refresh access token for deregister");
+    cloud_api_param_t *p = (cloud_api_param_t *)data;
+    free_api_param(p);
+    cloud_context_clear(ctx);
+    return;
+  }
+
+  // invoke in a delayed callback so, the cloud_api_param_t* structure allocated
+  // by cloud_deregister is deallocated before new one is allocated
+  cloud_reset_delayed_callback(data, cloud_deregister_refreshed_token_async, 0);
```

Comment:
this always add `cloud_deregister_refreshed_token_async` because data is always allocated